### PR TITLE
Split project tests

### PR DESCRIPTION
**Pull request by jakob-fritz**
Splitting tests from base-environment into separate jobs to increase parallelism and speed

### CODE REVIEW

**Comment by jakob-fritz** [comment]
@brownbaerchen is it on purpose, that the tests for monodomain are marked with "monodomain" instead of e.g. "base"? See all test-files in directory. E.g. [this one](https://github.com/Parallel-in-Time/pySDC/blob/376ab368848940ecd1a50da42ba584b083dd90f1/pySDC/tests/test_projects/test_monodomain/test_monodomain_convergence.py#L99)

I was surprised and assume, that those tests won't run, because currently only tests for base or other keywords will run.

However, we can also "invert" the decision and I run all tests in projects, that are not marked with any of the keywords, that need another environment-file

**Comment by jakob-fritz** [comment]
Ignore question above. I saw, that monodomain needs packages, that are not included in "base" environment. Therefore, I created a separate env and moved the monodomain tests, so that they don't run with the base-env (and fail)

**Comment by brownbaerchen** [comment]
The libpressio and monodomain tests need compiled code. That's why they're treated separately already and are not part of the matrix. You can just leave them as they are.

File: .github/workflows/ci_pipeline.yml
Comment (brownbaerchen) [inline .github/workflows/ci_pipeline.yml:121]: This runs all of the project tests in the base environment except those with mpi4py, fenics or petsc, yes? I think what @pancetta wanted is to allow for project specific environments.
What about using [`conda env update`](https://docs.conda.io/projects/conda/en/latest/commands/update.html) with `--file` option? This way we could maybe define a base environment and build project environments on top which only need to specify what is different to the base environment.

**Review by brownbaerchen — CHANGES_REQUESTED** [review]

File: .github/workflows/ci_pipeline.yml
Comment (jakob-fritz) [reply .github/workflows/ci_pipeline.yml:123]: Added this now. It uses `micromamba install` because the `update` really only updates packages that are already installed.

So all projects now got an additional file `env_update.yml` where pytest is installed. Pytest also is included in the base-env, but the files cannot be empty. As an alterantive another package could also be used if preferred.

File: .github/workflows/ci_pipeline.yml
Comment (jakob-fritz) [reply .github/workflows/ci_pipeline.yml]: Added `base` environment back to matrix, so that all base-tests from outside projects are now also tested again

**Comment by jakob-fritz** [comment]
@brownbaerchen do you know, why this test on [SDC_showdown](https://github.com/Parallel-in-Time/pySDC/actions/runs/8892081318/job/24415404605?pr=426) always fails?

I tries to convert a float into a matrix, but I don't see what i changed that triggered this error.

**Comment by brownbaerchen** [comment]
Cool! Can you add a short description of how to add a new project in the [contribution guide](https://github.com/Parallel-in-Time/pySDC/blob/master/CONTRIBUTING.md).
The steps are:
 - Add a folder in pySDC/projects
 - Add a non-empty environment update file called `env_update.yml` in the project directory
 - Add project name to the GitHub pipeline
 - Add tests to a directory called `tests` inside the project folder

That's it, right? I am wondering if we can forgo the need by the user to add the project name to the pipeline. Can we just run the tests for all directories in pySDC/projects except ones we specifically exclude?

**Comment by brownbaerchen** [comment]
> @brownbaerchen do you know, why this test on [SDC_showdown](https://github.com/Parallel-in-Time/pySDC/actions/runs/8892081318/job/24415404605?pr=426) always fails?
> 
> I tries to convert a float into a matrix, but I don't see what i changed that triggered this error.

Hm. I think @pancetta wrote this test and probably knows best. The petsc environment limits the version to `petsc4py<3.20`. Have you tried that?

**Comment by jakob-fritz** [comment]
> The petsc environment limits the version to `petsc4py<3.20`. Have you tried that?

Thank you. That fixed it.

**Comment by jakob-fritz** [comment]
> Cool! Can you add a short description of how to add a new project in the [contribution guide](https://github.com/Parallel-in-Time/pySDC/blob/master/CONTRIBUTING.md). The steps are:
> 
> * Add a folder in pySDC/projects
> * Add a non-empty environment update file called `env_update.yml` in the project directory
> * Add project name to the GitHub pipeline
> * Add tests to a directory called `tests` inside the project folder
> 
> That's it, right? I am wondering if we can forgo the need by the user to add the project name to the pipeline. Can we just run the tests for all directories in pySDC/projects except ones we specifically exclude?

That is right, with a one annotations. The environment-file must not only be non-empty, but also with the correct structure and in YAML. See any environment-file in the other projects.

I don't know a way how the list of projects could be done automatically in CI __before__ starting the jobs. Therefore, it is necessary to add the (new) own project to the list in the CI-file manually.

I read through the contribution guide and couldn't figure out, where these pieces of information fit best, i.e. where they are most likely to be found by the relevant people, when a new project is created. What is your view on the preferred location?

**Comment by pancetta** [comment]
I think I don't understand why we need the (empty or dummy) `env_update.yml` file. This is unclear to people writing projects and seems redundant? If we require an env-file for each project anyway, why not the full one? Then it's at least understandable what needs to be in there.

**Comment by brownbaerchen** [comment]
> I think I don't understand why we need the (empty or dummy) `env_update.yml` file. This is unclear to people writing projects and seems redundant? If we require an env-file for each project anyway, why not the full one? Then it's at least understandable what needs to be in there.

I guess the main reason we went with this is that there is a lot of pytest stuff that is annoying. How about doing both? We can specify all modules we need in the projects env files and use that to update the environment from the base environment in the pipeline such that we don’t need four obscure pytest installations in each project.

**Comment by jakob-fritz** [comment]
> I think I don't understand why we need the (empty or dummy) `env_update.yml` file. This is unclear to people writing projects and seems redundant? If we require an env-file for each project anyway, why not the full one? Then it's at least understandable what needs to be in there.

Of course, we can also go with 'only' the environment-file per project. The reason why I chose this approach is to avoid (or at least reduce) duplication of depedencies. If we go for a 'complete' environment-file per project, many dependencies will duplicate leading to an increased risk of diverging over time, when only a few environment-files are updated. Therefore, the idea was to have a common basis and putting the requirements that are specific for a project into this second file.

In addition to what @brownbaerchen said, we can also put the requirements that are only needed for testing into a separate env-file. Then they are installed in CI, but not when using pySDC locally.

Getting rid of the base-environment-file completely will be difficult, as we need some file for tests of the code, that is not part of any project.

**Comment by pancetta** [comment]
Why do you see the problem of divergence? Unless we don't pinpoint versions, there is no divergence, right? And if we had global requirements, these should go into the base file. And if a test fails to run, someone has to fix the environment (if that's the problem). Maybe I'm just not understanding the problem here?

**Comment by brownbaerchen** [comment]
> > I think I don't understand why we need the (empty or dummy) `env_update.yml` file. This is unclear to people writing projects and seems redundant? If we require an env-file for each project anyway, why not the full one? Then it's at least understandable what needs to be in there.
> 
> Of course, we can also go with 'only' the environment-file per project. The reason why I chose this approach is to avoid (or at least reduce) duplication of depedencies. If we go for a 'complete' environment-file per project, many dependencies will duplicate leading to an increased risk of diverging over time, when only a few environment-files are updated. Therefore, the idea was to have a common basis and putting the requirements that are specific for a project into this second file.
> 
> In addition to what @brownbaerchen said, we can also put the requirements that are only needed for testing into a separate env-file. Then they are installed in CI, but not when using pySDC locally.
> 
> Getting rid of the base-environment-file completely will be difficult, as we need some file for tests of the code, that is not part of any project.

I think it actually makes sense to decouple the base environment from the project environments because it is totally fine if they diverge. The reason we test the projects in the first place is that we want results in papers to be reproducible. If this requires limiting the version number of some dependency then that is totally fine. If the rest of the code runs with the most up-to-date version of the modules, then we should offer that in the base environment.

Let's do an environment for testing only and remove testing stuff from all environments and add it in by updating the environments. Are you fine with that @pancetta?
Then we can put everything in one matrix job again by using pytest tags for each project.
Can we somehow let the pipeline test all tags that were added to pytest instead of the user having to add the tag to the pipeline manually?

**Comment by jakob-fritz** [comment]
I splitted/decoupled the environments of the projects from the base-environment. Now there is one environment for each project, where only the needed dependencies are included. Because of this change, I also changed the name to `environment.yml`, which is also the default name for conda/mamba. Another environment-file now contains all the pytest-related dependencies, that are only needed for testing (so not for productive use).
The base-environment still exists and is used for the tests that are not part of any project.

Using the project-environments to simply update the base-environment is the way, the files were designed before my changes today. There, the base-env included all dependencies (incl. pytest-stuff) for the majority of the code and the env-files in the projects contained additional dependencies if needed. If no additional dependencies were needed, I included a dummy-dependency (I chose pytest to make clear it's no dependency for the actual pySDC-code), as the files needs to exist and follow a valid structure. But again, I changed this structure today to self-sustained environment-files for the projects. Otherwise, I can also revert those changes.

Combining the project-tests and the other tests into a single matrix might be difficult, as the paths follow different patterns. Therefore, conditionals need to be included in the job, when the matrix is combined. This is doable, but I wanted to do those steps after each other.

**Comment by jakob-fritz** [comment]
There is still the issue, that the tests in parallelSDC (in the file [test_preconditioner_MPI](<https://github.com/jakob-fritz/pySDC/blob/split_project_tests/pySDC/projects/parallelSDC/tests/test_preconditioner_MPI.py>)) still take very long (over 1h) to complete. I don't know why, it takes that long. Do you, @brownbaerchen @pancetta have any idea how this can be fixed, or if these jobs should be omitted?

**Comment by pancetta** [comment]
> There is still the issue, that the tests in parallelSDC (in the file [test_preconditioner_MPI](https://github.com/jakob-fritz/pySDC/blob/split_project_tests/pySDC/projects/parallelSDC/tests/test_preconditioner_MPI.py)) still take very long (over 1h) to complete. I don't know why, it takes that long. Do you, @brownbaerchen @pancetta have any idea how this can be fixed, or if these jobs should be omitted?

But they didn't take that long before, right?

File: .github/workflows/ci_pipeline.yml
Comment (brownbaerchen) [inline .github/workflows/ci_pipeline.yml:47]: Don't forget to revert this once we're ready to merge.

File: .github/workflows/ci_pipeline.yml
Comment (brownbaerchen) [inline .github/workflows/ci_pipeline.yml:114]: @pancetta also wants all python versions tested here. Btw: Did you test 3.11 and 3.12? Especially with the non-project tests?

File: .github/workflows/ci_pipeline.yml
Comment (brownbaerchen) [inline .github/workflows/ci_pipeline.yml:140]: Did you check that the website looks good? The plots are put into a relative path, which is likely wrong when we move the tests. Sorry that I only remember this now. Maybe we better put the tests back to where they were. Then we can use the same matrix for everything. But the matrix we can do either way, I think, when we use the pytest markers and just test everything in `pySDC/tests` and `pySDC/tests/projects`.

File: .github/workflows/ci_pipeline.yml
Comment (brownbaerchen) [inline .github/workflows/ci_pipeline.yml:183]: Let's not remove the pytest marker

File: .github/workflows/ci_pipeline.yml
Comment (brownbaerchen) [inline .github/workflows/ci_pipeline.yml:231]: Also here I'd like to keep the marker

File: .github/workflows/ci_pipeline.yml
Comment (brownbaerchen) [inline .github/workflows/ci_pipeline.yml:306]: Are we sure this is enough? Can you try running this job locally and see if the website looks good?

File: docs/contrib/06_new_project.md
Comment (brownbaerchen) [inline docs/contrib/06_new_project.md:4]: When you do a project with pySDC we appreciate it if you merge it back to the main repository. We are committed to keeping your work reproducible and prevent it from fading into oblivion.
To that end, please write extensive tests for your code and add them to the project. See the [contribution guide](../../../CONTRIBUTING.md) for general advice on testing etc. This guide will detail only how to add a project to pySDC.

File: docs/contrib/06_new_project.md
Comment (brownbaerchen) [inline docs/contrib/06_new_project.md:13]: The testing pipeline uses [micromamba](https://mamba.readthedocs.io/en/latest/user_guide/micromamba.html) and requires an environment file for setup.

File: docs/contrib/06_new_project.md
Comment (brownbaerchen) [inline docs/contrib/06_new_project.md:15]: `env_update.yml` is outdated

File: docs/contrib/06_new_project.md
Comment (brownbaerchen) [inline docs/contrib/06_new_project.md:33]: We need to remember to put whatever we settle on in the end here.

File: docs/contrib/06_new_project.md
Comment (brownbaerchen) [inline docs/contrib/06_new_project.md:35]: ## Add the project to the continuous integration pipeline

File: docs/contrib/06_new_project.md
Comment (brownbaerchen) [inline docs/contrib/06_new_project.md:40]: Again, we need to make sure that this is correct in the end. If we decide to use the. pytest markers (I vote we do) we need to detail here how to add you own.

Comment (brownbaerchen) [inline docs/contrib/06_new_project.md:40]: @pancetta, do you want to add a few lines on how to obtain a release of pySDC to go along with a publication?

File: etc/environment-mpi4py.yml
Comment (brownbaerchen) [inline etc/environment-mpi4py.yml:15]: @pancetta why was `pytest-order` missing here? Are we even using that?

File: etc/environment-tests.yml
Comment (brownbaerchen) [inline etc/environment-tests.yml:11]: Does it make sene to do a special environment for post processing such we don't need to install `sphinx` in every environment? I don't know how long this takes. But if we can shave a few seconds off every single pipeline, that might be worth doing.

File: pySDC/projects/parallelSDC/environment.yml
Comment (brownbaerchen) [inline pySDC/projects/parallelSDC/environment.yml:10]: Let's put exactly the base environment here and see if the test times change.

**Review by brownbaerchen — CHANGES_REQUESTED** [review]

File: docs/contrib/06_new_project.md
Comment (pancetta) [reply docs/contrib/06_new_project.md:40]: Something like:
```
If you project is published and you need a dedicated pySDC version with a DOI, please get in touch with us and/or open a new issue. We will help you with this as soon as possible. Note that a final DOI is usually only necessary once a paper is accepted and the final proofs are due. We strongly encourage to describe and cite the current version of pySDC already during initial submission, though.
```

File: etc/environment-mpi4py.yml
Comment (pancetta) [reply etc/environment-mpi4py.yml:15]: I.... don't know.

File: .github/workflows/ci_pipeline.yml
Comment (brownbaerchen) [reply .github/workflows/ci_pipeline.yml:114]: I forgot what the issue was. That's why I wanted to test everything with everything and see if the base stuff at least works with the latest python. That would be nice because I would like to transition to a newer version on my laptop. On the cluster I have been using 3.11 with no issues. Actually, the HPC tests are probably all run with 3.11, maybe..?

File: etc/environment-mpi4py.yml
Comment (jakob-fritz) [reply etc/environment-mpi4py.yml:15]: We use it only in one project (TOMS). As the tests there are marked with "base", there is no need for pytest-order there. For simplicity I included it in environment-tests anyway, as one job needs it.

File: .github/workflows/ci_pipeline.yml
Comment (jakob-fritz) [reply .github/workflows/ci_pipeline.yml:114]: No, I did not test more python versions. This reduction on python-versions is also only for current development, as the CI is faster with fewer jobs and finds most of the errors also with only a single python-version.

It is not meant to stay like that for merging.

File: .github/workflows/ci_pipeline.yml
Comment (jakob-fritz) [reply .github/workflows/ci_pipeline.yml:183]: We can also use the markers. I chose this change to make sure to include all tests there; also ones, where the marker may be missing. But We can also include the requirement for a marker there, if you prefer that.

File: .github/workflows/ci_pipeline.yml
Comment (jakob-fritz) [reply .github/workflows/ci_pipeline.yml:231]: Same as above. I chose to remove the marker, as the path limits the tests already thematically. Therefore, the marker was an additional potential point of failure (skipping tests instead of running them). But also here: We/I can also include that marker again.

File: docs/contrib/06_new_project.md
Comment (jakob-fritz) [reply docs/contrib/06_new_project.md:4]: Changes will be included in the next commit.

File: docs/contrib/06_new_project.md
Comment (jakob-fritz) [reply docs/contrib/06_new_project.md:13]: Will also be included in the next commit

File: docs/contrib/06_new_project.md
Comment (jakob-fritz) [reply docs/contrib/06_new_project.md:15]: Name of the file is updated in the next commit.

File: docs/contrib/06_new_project.md
Comment (jakob-fritz) [reply docs/contrib/06_new_project.md:35]: Headline updated in the next commit

File: pySDC/projects/parallelSDC/environment.yml
Comment (jakob-fritz) [reply pySDC/projects/parallelSDC/environment.yml:7]: I just gave it a try with an environment that is similar (not the same) to the mpi4py-environment. The local results looked promising. Let's see how CI will behave.

File: .github/workflows/ci_pipeline.yml
Comment (brownbaerchen) [reply .github/workflows/ci_pipeline.yml:183]: Yes please. The marker marks tests that should be run in the docker container. Let's use it like it's intended.

File: etc/environment-mpi4py.yml
Comment (pancetta) [reply etc/environment-mpi4py.yml:15]: Not sure why I did this [here](https://github.com/Parallel-in-Time/pySDC/blob/master/pySDC/tests/test_projects/test_TOMS/test_AllenCahn_contracting_circle.py), but I guess by merging the two tests we could get rid of it.

Comment (brownbaerchen) [reply etc/environment-mpi4py.yml:15]: Actually, what you did there is quite nice! I suggest we keep this module and use it more often for plotting for the website. Personally, I really like splitting the tests as much as possible so you get fine-grained feedback, but then you cannot plot all of the results in a single plot without this module.

Comment (pancetta) [reply etc/environment-mpi4py.yml:15]: Fine by me!

File: pySDC/projects/parallelSDC/environment.yml
Comment (jakob-fritz) [reply pySDC/projects/parallelSDC/environment.yml:7]: As the jobs for parallelSDC need only 10-20 minutes (the first one and the reloaded one respectively), I assume, we can close this conversation. If you think otherwise, please reopen it or open a new one.

File: docs/contrib/06_new_project.md
Comment (brownbaerchen) [reply docs/contrib/06_new_project.md:47]: I like the markers because they assign the tests to the respective environments in a very explicit way. But I can see that we disagree on this issue. @pancetta, please note your preference and we'll go with that.

File: .github/workflows/ci_pipeline.yml
Comment (jakob-fritz) [reply .github/workflows/ci_pipeline.yml]: Reverted in eebbdd3b89a71f3c3063efccfab926abde748fea

File: .github/workflows/ci_pipeline.yml
Comment (jakob-fritz) [reply .github/workflows/ci_pipeline.yml]: Reverted in eebbdd3b89a71f3c3063efccfab926abde748fea and also added more python-versions for testing (3.11 & 3.12)

File: .github/workflows/ci_pipeline.yml
Comment (jakob-fritz) [reply .github/workflows/ci_pipeline.yml]: Readded the marker in b858c33c7fdec93b049845f5232f5b817c127c27

File: docs/contrib/06_new_project.md
Comment (pancetta) [reply docs/contrib/06_new_project.md:49]: The markers were used originally because we wanted to filter the tests by a certain property. Adding new ones is pretty clumsy and forgetting to set a marker was also easily possible. So, if we don't need markers, we should get rid of them. But if we need to do other filter magic instead, we should keep them. So, what would happen if we drop markers?

File: .github/workflows/ci_pipeline.yml
Comment (jakob-fritz) [reply .github/workflows/ci_pipeline.yml]: Readded the marker in b858c33c7fdec93b049845f5232f5b817c127c27

File: docs/contrib/06_new_project.md
Comment (pancetta) [reply docs/contrib/06_new_project.md:49]: In any case, we should not mix markers and non-markers. If we decide to use markers, we should only use markers. If we decide to get rid of markers, we should get fully rid of them.

Comment (jakob-fritz) [reply docs/contrib/06_new_project.md:49]: Added the text proposed by @pancetta in commit 1a4b58f591c1e0a0f1fe1e22922c3f16a81f046c and also added a headline as it is not directly related to the CI-Pipeline

Comment (brownbaerchen) [reply docs/contrib/06_new_project.md:49]: Jakob currently set things up so that the project tests are in the project folders and the pipeline just runs all tests in a project folder with the environment file in the same folder. The remaining tests are still in the separate tests directory and use the markers `base`, `mpi4py`, ... 
So then you don't need to add markers when adding a new project. Fine. I agree that it's a bit simpler ;)

File: docs/contrib/06_new_project.md
Comment (brownbaerchen) [inline docs/contrib/06_new_project.md:10]: I think we don't need the above two lines anymore. My goal was to elaborate on these points with the above text.

Comment (jakob-fritz) [reply docs/contrib/06_new_project.md:10]: Fine. Will be removed in next commit

**Comment by jakob-fritz** [comment]
One job fails for python 3.12. The issue arises for higher number of Threads (5 and 8) but does not arise for older versions of python <=3.11 (or lower numbers of threads).
See log [here](<https://github.com/Parallel-in-Time/pySDC/actions/runs/8936665061/job/24547469281?pr=426>)
This is not a one-time-issue, but occured at least twice for two pushes.

The issue seems to come from using a different MPI. When comparing the installation logs, in Python 3.11 impi (sometimes listed as impi_rt) is used. In Python 3.12 OpenMPI is installed instead. Maybe these two packages have different default options (maybe for oversubscription or the likes?)

Do you @brownbaerchen want to have a look there, or shall I dig in, but that will take more time, as I haven't worked with the pySDC code itself until now.

**Comment by pancetta** [comment]
Looks like a oversubscription issue which is indeed fixed by either specifying a parameter during `mpirun` or by using another MPI version (I think MPICH does not have this problem).

**Comment by jakob-fritz** [comment]
> Looks like a oversubscription issue which is indeed fixed by either specifying a parameter during `mpirun` or by using another MPI version (I think MPICH does not have this problem).

Yes, adding mpich in the requirements solved it. Thank you @pancetta for the pointer!

**Comment by brownbaerchen** [comment]
Looking good! Testing the past 5 versions is perhaps a bit excessive. Which ones do you want to keep @pancetta?

**Comment by pancetta** [comment]
Looking [here](https://devguide.python.org/versions/), v3.8 to v3.12 would be great to have. I'd be fine dropping the oldest one, though.

**Comment by pancetta** [comment]
So, runtime of the tests went up to 1h.. is that intended?

**Comment by jakob-fritz** [comment]
No it is not. This is because of matplotlib. When I limit the version, the duration goes down to 10-20 minutes. But fails for Python 3.11 and Python 3.12., because those older versions of matplotlib are not compatible with the later python versions.

I am currently at it and try to find where this issue with matplotlib comes from

**Comment by pancetta** [comment]
Ahhh, yes, I recall that I did not pursue 3.12 because Matplotlib was causing trouble... right.

**Comment by brownbaerchen** [comment]
Hm. That's annoying. What about testing projects only up to 3.10 and the rest with everything? Since only the projects do plots (right? Tutorials..?) we don't need to worry about matplotlib for the rest.

**Comment by jakob-fritz** [comment]
I added a statement to skip the combination of parallelSDC and Python 3.11 or Python 3.12. So all other projects are tested with 5 versions of python and the project parallelSDC is tested with "only" 3 versions of python. By this, we can cover as much as possible for the other projects, while keeping the version-limits for matplotlib

File: pySDC/projects/parallelSDC/tests/test_preconditioner_MPI.py
Comment (brownbaerchen) [inline pySDC/projects/parallelSDC/tests/test_preconditioner_MPI.py:1]: Why is this file marked as new instead of moved in the diff? Did you check in the "removed" version in the original directory? Probably some GitHub bug...

**Comment by brownbaerchen** [comment]
By chance, the relative paths for the plots still work out. So the website and coverage report look like before. Time is down to ~30 mins from ~45. More Python versions including the latest 3.12 are tested.

To get time down further, we would have to rewrite some tests, mainly in the DAE, resilience and the parallelSDC_reloaded projects. But as they generate some plots, this may be difficult and definitely shouldn't be part of this PR.

I am happy now. Any more requests, @pancetta?

**Review by brownbaerchen — APPROVED** [review]

File: docs/contrib/06_new_project.md
Comment (jakob-fritz) [reply docs/contrib/06_new_project.md:46]: Just to avoid confusion: We can get rid of all markers within the tests for projects. But currently we still use pytest-markers for all tests outside of projects. I wanted to explicitly state that as it slightly contradicts the statement of @pancetta about using markers everywhere or nowhere.
Is that fine with you, that we use them nowhere within the project-tests and everywhere outside of project-tests, @pancetta ?
I don't see a good way to get rid of the markers for the tests outside of projects, but don't think we need them within project-tests.

**Comment by pancetta** [comment]
Ok, let me try to understand this: each project now has a separate environment file, yes? And then there are environment files for the tutorials (and other stuff?) in the `env` folder? Does it make sense to keep those separate? Why not do this consistently?

File: docs/contrib/06_new_project.md
Comment (pancetta) [reply docs/contrib/06_new_project.md:46]: Yeah, we did introduce inconsistencies here, too (see also my other comment below).

**Comment by jakob-fritz** [comment]
Yes, you got this right, @pancetta. The environment-file for each project is located in the directory of that project. Furthermore, there are more general environment-files (e.g. for tutorial, the tests outside projects and with testing-dependencies in general) in the `etc` directory. We could move the env-files of the projects also there, but then the projects are less self-contained. If we keep the environment-files in the project-dirs, the environment-files are in two places (conceptually; of course, many more, since the files are in the project-dirs).

**Comment by pancetta** [comment]
Alright, there is probably no single great solution to this. I really appreciate that the projects are now more self-contained and that the tests are faster/more modular. Any famous last words here, @brownbaerchen @jakob-fritz ?

**Comment by jakob-fritz** [comment]
> Alright, there is probably no single great solution to this. I really appreciate that the projects are now more self-contained and that the tests are faster/more modular. Any famous last words here, @brownbaerchen @jakob-fritz ?

Not last words: Please don't merge yet. I am adding dependencies for the documentation (did not raise errors, but only warnings in the logs, I saw this morning).

**Comment by jakob-fritz** [comment]
Now, the documentation has all needed dependencies installed.
I just pushed a minor change, where I removed print-statements in the CI, that I added earlier today for debugging.

Once this is done, I am fine with squash and merge

File: .github/workflows/ci_pipeline.yml
Comment (jakob-fritz) [reply .github/workflows/ci_pipeline.yml]: Now the environment for documentation has all needed dependencies

File: docs/contrib/06_new_project.md
Comment (jakob-fritz) [reply docs/contrib/06_new_project.md]: Should work this way, as we don't need pytest-markers and only put the files in the correct spot.

**Pull request merged**
